src/consensus/dkg.rs

```diff
@@ -228,37 +266,32 @@ impl DkgVoter {
                 let mut commands = vec![];
 
                 for message in messages {
-                    commands.extend(session.broadcast(our_name, dkg_key, message));
+                    commands.extend(session.broadcast(dkg_key, message));
                 }
                 commands.push(session.reset_timer());
-                commands.extend(self.check());
+                commands.extend(self.check(our_name));
                 commands
             }
             Err(error) => {
                 trace!("DKG for {} failed: {}", elders_info, error);
 
                 let elders_info = session.elders_info.take().unwrap();
 
-                self.participant = None;
-
-                vec![DkgCommand::HandleParticipationResult {
-                    dkg_key,
-                    elders_info,
-                    result: Err(()),
-                }]
+                // Restart on failure.
+                self.start(our_name, dkg_key.retry(), elders_info)
```

Comment:
could cause different node using different version of dkg_key (of the same eldres_info),
depends on the different times of time_out experienced ?

Reply:
Yes, but that's on purpose to avoid the messages from different DKG session (the original one and the restarted one) to mix.

Reply:
I am concerned this may cause a deadlock, say some is always at version `n`, meanwhile some alwasy at version `n+1`, and they keeps growing with the diff of `1` always exists.

Reply:
Hm, I guess that's not impossible. So say a node A detects a failure and switches to version `n+1` but the other nodes are taking longer to detect the failure themselves, so node A fails again and switches to `n+2` and the other nodes switch to `n+1` only after that, this never catching up.

There is also another issue I realized yesterday, although it's not as serious. Say we start DKG with the participants A, B, C, D, E. Then some time later, but before the first DKG completes, we start another DKG with A, B, C, F, G. Then D and E will still think they are on the first DKG and will keep trying to progress it, fail, restart and so on, basically forever.

OK, I'll try to think about how to resolve this. Will probably have to revert part of this PR.